320 introduce houdini generic loader hdas with examples for users (#321)
ynput/ayon-documentation: website/docs/addon_houdini_artist_loader_hdas.md

---
id: addon_houdini_artist_loader_hdas
title: AYON Loader HDAs
sidebar_label: Loader HDAs
description: AYON Loader HDAs Documentation.
toc_max_heading_level: 5
---

import ReactMarkdown from "react-markdown";
import versions from '@site/docs/assets/json/Ayon_addons_version.json'

<ReactMarkdown>
{versions.Houdini_Badge}
</ReactMarkdown>

## Introduction

> AYON Loader HDAs act as controllers that allows you to select AYON products to load into your scene.

Reviewer: ```suggestion
AYON Loader HDAs act as controllers that allow you to select AYON products to load into your scene.
```

Author: https://github.com/ynput/ayon-documentation/pull/321/commits/e44be9caa7b802f754c9c3c135c60f3a30aa4d30

- Generic Loader as its name implies it's generic, it loads any AYON product as a filepath.
- Lop Import (Load Asset): load USD assets. 
- Load Shot: Load USD shots.

| Generic Loader |
|--|
| Currently, Available in `OBJs`, `SOPs` and `LOPs`. ![](assets/houdini/artist/hda_generic_loader.png) |

| LOP Import | Load Shot |
|--|--|
| ![](assets/houdini/artist/hda_lop_import.png) | ![](assets/houdini/artist/hda_load_shot.png) |

## Access Loader HDAs

### TAB Menu
All of these nodes are accessible in TAB menu.

| LOPs | OBJs | SOPs |
|--|--|--|
| ![](assets/houdini/artist/hda_lop_tab_menu.png) | ![](assets/houdini/artist/hda_obj_tab_menu.png) | ![](assets/houdini/artist/hda_sop_tab_menu.png) |

### Loader Tool

> Some loaders depends on these nodes to perform loading.

Reviewer: ```suggestion
Some loaders depend on these nodes to perform loading.
```

Author: https://github.com/ynput/ayon-documentation/pull/321/commits/e44be9caa7b802f754c9c3c135c60f3a30aa4d30

- **File**: Shows the file of the selected product. It automatically applies `$F` and `<UDIM>` tokens.
- **Use AYON Entity URI**: When enabled the File parameter will turn into an AYON URI.
- **Show Entity Thumbnail**: When enabled, it shows the selected product's thumbnail and show extra parameters for adjusting the image in network view.
- **Size** & **Offset**: They are used to adjust the displayed thumbnail in network view.
- **Update Thumbnail**: Click to manually update the thumbnail, in some cases auto update doesn't work, so you'd need to force update.
- **Show Pipeline Parms**: It shows extra pipeline parms like `Representation ID`, It's mostly used for debugging and automations.
- **Nodes Referencing File**: It shows nodes referencing the `File` parameter.

#### How to use Generic Loader

Select project and folder path
Select product, version and representation.
Supports expressions and context options. more examples check

### Lop Import 

#### Overview
It uses a combination of [Generic Loader](#generic-loader) and [Reference](https://www.sidefx.com/docs/houdini/nodes/lop/reference.html) node

![](assets/houdini/artist/hda_lop_import_parameters.png)

#### Parameters

> It shares the same parameters as generic loader except for `Nodes referencing file` and also add few more parameters.

Reviewer: ```suggestion
It shares the same parameters as generic loader except for `Nodes referencing file` and also adds few more parameters.
```

Author: https://github.com/ynput/ayon-documentation/pull/321/commits/e44be9caa7b802f754c9c3c135c60f3a30aa4d30

It also exposes some parameters of [Reference](https://www.sidefx.com/docs/houdini/nodes/lop/reference.html) node. 

- **Reload Files**
- **Primitive Root** & **Presets**
- **Reference Type**
- **Make Instanceable**
- **Reference Primitive**
- **Reference Primitive Path**
- **Time Offset (in Frames)**
- **Time Scale**

#### How to use LOP Import

> - Select products works the same as [generic loader](#how-to-use-generic-loader).

Reviewer: ```suggestion
- Select products works the same with [generic loader](#how-to-use-generic-loader).
```

Author: https://github.com/ynput/ayon-documentation/pull/321/commits/e44be9caa7b802f754c9c3c135c60f3a30aa4d30

> - Set `Primitive Root`, the `Presets` menu provides provides default values for building assets and shots.

Reviewer: ```suggestion
- Set `Primitive Root`, the `Presets` menu provides default values for building assets and shots.
```

Author: https://github.com/ynput/ayon-documentation/pull/321/commits/e44be9caa7b802f754c9c3c135c60f3a30aa4d30

### Load Shot

#### Overview
It uses a combination of [Generic Loader](#generic-loader) and [Sublayer](https://www.sidefx.com/docs/houdini/nodes/lop/sublayer.html) node.

![](assets/houdini/artist/hda_load_shot_parameters.png)

#### Parameters

> It share the same parameters as generic loader except for `Nodes referencing file` and also add few more parameters.

Reviewer: ```suggestion
It share the same parameters with generic loader except for `Nodes referencing file` and also add few more parameters.
```

Author: https://github.com/ynput/ayon-documentation/pull/321/commits/e44be9caa7b802f754c9c3c135c60f3a30aa4d30

- **Reload Files**
- **Mute Layer**
- **Time Offset (in Frames)**
- **Time Scale**

#### How to use Load Shot

> - Select products works the same as [generic loader](#how-to-use-generic-loader).

Reviewer: ```suggestion
- Select products work the same with [generic loader](#how-to-use-generic-loader).
```

Author: https://github.com/ynput/ayon-documentation/pull/321/commits/e44be9caa7b802f754c9c3c135c60f3a30aa4d30

> - Keep in mind, it adds loaded product as a sublayer.

Reviewer: ```suggestion
- Keep in mind, it adds the loaded product as a sublayer.
```

Author: https://github.com/ynput/ayon-documentation/pull/321/commits/e44be9caa7b802f754c9c3c135c60f3a30aa4d30

## Example Usages
These are examples from various discussions on our community.
You can find here some inspiration.

> Although They might not fulfill your needs at your studios but they can serve as good starting point.

Reviewer: ```suggestion
Although They might not fulfill your needs at your studios, they can serve as good starting point.
```

Author: https://github.com/ynput/ayon-documentation/pull/321/commits/e44be9caa7b802f754c9c3c135c60f3a30aa4d30

:::tip
Feel free to share us your examples on [Ynput Community Forums](https://community.ynput.io/).
:::

### Context Options and MultiShots

You can use 'context options' entry to select from - where the user is capable of selecting any folder in the project.

> e.g. you can create a menu that includes all folder paths that has tasks in the current project.

Reviewer: ```suggestion
e.g. you can create a menu that includes all folder paths that have tasks in the current project.
```

Author: https://github.com/ynput/ayon-documentation/pull/321/commits/e44be9caa7b802f754c9c3c135c60f3a30aa4d30

![Example 1](assets/houdini/artist/hda_example_1_expressions.png)

Next, I loaded them into my scene using loader HDAs in a loop with expressions.

- **LOP, expressions, and context options:**
  ```
  pointcacherock_`@ITERATIONVALUE+1`
  ```
  ![LOP Expressions](assets/houdini/artist/hda_expressions_lop.png)
- **SOP, expressions, and detail attribute:**
  ```
  pointcacherock_`detail("../foreach_count","iteration", 0)+1`
  ```
  ![SOP Expressions](assets/houdini/artist/hda_expressions_sop.png)

### Build Your Own HDAs Using Generic Loader

One of the main uses of loader HDAs is to enable users to incorporate them into their own HDAs or build new HDAs on top of them.
My custom HDA example is based on using unified naming conventions for materials in my project. <br/>
In my HDA, I just need to select the material folder path within my library project and specify the version. I can also set the material prim path prefix.

> Download my example custom HDA: [Expr_random_hda_Material_Loader_v001.hda](assets/houdini/artist/Expr_random_hda_Material_Loader_v001.hda).

![Custom Loader](assets/houdini/artist/hda_custom_loader.png)

:::info

> For this example, textures were published with the product type `image` and variant name `materialName_AOV` using [Tray Publisher](addon_traypublisher_artist.md).

Reviewer: ```suggestion
For this example, textures are published with the product type `image` and variant name `materialName_AOV` using [Tray Publisher](addon_traypublisher_artist.md).
```
